### ISSUE TITLE
Python USDT probes in GKE cluster

### ISSUE THREAD

**Issue by EliasLindkvist** [iovisor/bcc#3945]
Hi,

I'm trying to trace a Python application running in a GKE cluster and am having problems with enabling the USDT probe.

I've built Docker images which works when I run them locally with the Docker CLI, but when I try using the same images in kubernetes it fails to enable the probe.

I've ran `tplist` on the PID of the application and got the following result:

```
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:function__return
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:function__entry
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:line
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:import__find__load__start
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:import__find__load__done
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:audit
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:gc__start
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:gc__done
```

So the probes are there, and I've tried a bunch of different names for the probe with the `USDT.enable_probe` function.

```
function__entry
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0 python:function__entry
/proc/29/root/usr/local/lib/libpython3.8d.so.1.0:function__entry
/usr/bin/python3:function__entry
/usr/bin/python3.8:function__entry
/usr/bin/python:function__entry
/proc/29/root/usr/bin/python3.8:function__entry
/proc/29/root/usr/bin/python3:function__entry
/proc/29/root/usr/bin/python:function__entry
/proc/29/root/usr/local/bin/python3.8:function__entry
/proc/29/root/usr/local/bin/python3:function__entry
/proc/29/root/usr/local/bin/python:function__entry
```

Nothing seems to work. The error is not very descriptive, so I'm not sure what the problem is. Does anyone have any idea of what might help solve this?

**Comment by chenhengqi** [comment]
> The error is not very descriptive, so I'm not sure what the problem is.

Could you post the error message ?

**Comment by chenhengqi** [comment]
Do you run your BPF program with sudo ?

**Comment by EliasLindkvist** [comment]
> Do you run your BPF program with sudo ?
Yes, well since both my application and my BPF program is ran in containers I use the equivalent of sudo.
I have two images, 1 with my application and one with my BCC script.

I can run them locally using docker as such:
```
docker run --rm --name app -p 8000:8000 django-app:v1
docker run --rm --name trace --privileged --pid container:app python-calls:v1 $(docker exec -ti app pidof python -o 1)
```
With piviliged mode for the BPF container and sharing PID namespace.

I've tried using the exact same images with kubernetes and once again setting privileged mode and such. I found [this article](https://containerjournal.com/kubecon-na-2021/deploying-and-running-bcc-in-your-kubernetes-cluster/) which was the only source on using BCC in kubernetes I could find. So I've tried the following in my yaml:
```
securityContext:
  privileged: true
  runAsUser: 0
  capabilities:
    add:
      - SYS_ADMIN
```
I don't know if anything more is needed for full privileges, especially in GKE.

**Comment by chenhengqi** [comment]
Please refer to #3863, may be it helps.

**Comment by EliasLindkvist** [comment]
I saw that but it did not help me progress.

My USDT object finds all the probes:
```
usdt = USDT(pid=args.pid)
for probe in usdt.enumerate_probes():
    print(probe.name.decode())

function__return
function__entry
line
import__find__load__start
import__find__load__done
audit
gc__start
gc__done
```

The problem is enabling them, and I'm not sure what the correct path should be.
The USDT probe gives the binary path
`/proc/78309/root/usr/local/lib/libpython3.8d.so.1.0`

But this fails
`usdt.enable_probe('/proc/78309/root/usr/local/lib/libpython3.8d.so.1.0:function__entry')`

**Comment by chenhengqi** [comment]
`usdt.enable_probe` expects `${usdt_provider}:${usdt_name}`.

Could you please run `readelf -n /path/to/binary` and post the output here ?

**Comment by EliasLindkvist** [comment]
This is the readelf output 
```
root@gke-thesis-cluster-default-pool-5d2accf9-ps2g:/usr/sbin# readelf -n /proc/51115/root/usr/local/lib/libpython3.10d.so.1.0

Displaying notes found in: .note.gnu.build-id
  Owner                Data size 	Description
  GNU                  0x00000014	NT_GNU_BUILD_ID (unique build ID bitstring)
    Build ID: 2847651c8dec4c3368856efff8f42e75e9650fab

Displaying notes found in: .note.stapsdt
  Owner                Data size 	Description
  stapsdt              0x00000033	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: audit
    Location: 0x0000000000065789, Base: 0x0000000000361e18, Semaphore: 0x0000000000403756
    Arguments: 8@%r13 8@%r14
  stapsdt              0x00000031	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: gc__start
    Location: 0x0000000000076e44, Base: 0x0000000000361e18, Semaphore: 0x000000000040374e
    Arguments: -4@%ebp
  stapsdt              0x00000030	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: gc__done
    Location: 0x0000000000077257, Base: 0x0000000000361e18, Semaphore: 0x0000000000403750
    Arguments: -8@%r11
  stapsdt              0x00000040	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: import__find__load__start
    Location: 0x00000000000bd1c8, Base: 0x0000000000361e18, Semaphore: 0x0000000000403752
    Arguments: 8@%rax
  stapsdt              0x00000047	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: import__find__load__done
    Location: 0x00000000000bd1e1, Base: 0x0000000000361e18, Semaphore: 0x0000000000403754
    Arguments: 8@%rax -4@%r8d
  stapsdt              0x00000046	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: function__return
    Location: 0x00000000002673dd, Base: 0x0000000000361e18, Semaphore: 0x0000000000403742
    Arguments: 8@%rbp 8@%r12 -4@%eax
  stapsdt              0x00000045	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: function__entry
    Location: 0x0000000000267417, Base: 0x0000000000361e18, Semaphore: 0x0000000000403740
    Arguments: 8@%rbp 8@%r12 -4@%eax
  stapsdt              0x0000003a	NT_STAPSDT (SystemTap probe descriptors)
    Provider: python
    Name: line
    Location: 0x0000000000267495, Base: 0x0000000000361e18, Semaphore: 0x000000000040374c
    Arguments: 8@%r12 8@%rax -4@%ebp

```

**Comment by EliasLindkvist** [comment]
Yes, it does not work.

**Comment by chenhengqi** [comment]
Can you post a reproducer here ?

**Comment by chenhengqi** [comment]
I can reproduce this locally, need some investigation.

**Comment by EliasLindkvist** [comment]
> Can you post a reproducer here ?

Its hard for me to post something you can just use directly because there is a lot of steps but I can try to explain the process I took.

I first needed to modify the python parent images found [here](https://github.com/docker-library/python) to use the --with-dtrace tag. (Following [this](https://blog.0x74696d.com/posts/debugging-python-containers-in-production/) blog post).

My tracing image use Ubuntu as a base and clone and compile BCC like [this article](https://containerjournal.com/kubecon-na-2021/deploying-and-running-bcc-in-your-kubernetes-cluster/) explains. I also use the initContainer and mount volumes as the article show. 

For Kubernetes I've tried both using two containers in the same pod with `shareProcessNamespace: true`, and two separate pods with `hostPID: true`, so that the tracer can find and attach to the application. The tracing container use the securityContext I described above. All of this run on a standard GKE cluster.

My tracer fetches the PID using `pgrep` and the creates a `bcc.USDT` object. It is then the `usdt.enable_probe` call that fails to enable the specified probe. Here as I've explained I have tried any imaginable provider name for the probe. 

(I have also tested creating a `bcc.BPF` object with my BPF program which succeeds, so I know the linux headers are on the GKE nodes.)

**Comment by chenhengqi** [comment]
```python
#!/usr/bin/python3
from bcc import BPF, USDT

bpf_prog = '''
int trace(struct pt_regs *ctx)
{
	bpf_trace_printk("function entry\\n");
	return 0;
}
'''

pid = 2761202
path = '/proc/2761202/root/usr/bin/python3.9'
usdt = USDT(path=path, pid=pid)
usdt.enable_probe('python:function__entry', 'trace')
b = BPF(text=bpf_prog, usdt_contexts=[usdt])
b.trace_print()
```

This script works for me.

I pulled the `python:3.10` Docker image, it contains a `python3.9` binary, which has `USDT` enabled by default.

**Comment by EliasLindkvist** [comment]
Still does not work for me. I can see the probes in the `python3.9` binary but same issue with failing to enable. Maybe it is some weird privilege issue with GKE?

**Comment by EliasLindkvist** [comment]
I've managed to find a fix which makes it work in GKE.

Changing the node image from the default `Container-Optimized OS with containerd (cos_containerd)` to `Ubuntu with Docker (ubuntu)` works. I can now enable the probe with simply `function__entry` (no provider). (I tried `Container-Optimized OS with Docker (cos)` last week but that did not work). 

Not sure why this is the case, but I have no time to further examine the issue.
